src/ln/channelmanager.rs

```diff
@@ -2077,152 +2121,92 @@ impl ChannelManager {
+								fail_channel_update = Some(msgs::HTLCFailChannelUpdate::NodeFailure { node_id: route_hop.pubkey, is_permanent: error_code & PERM == PERM });
+							}
+							else if error_code & PERM == PERM {
+								fail_channel_update = if payment_failed {None} else {Some(msgs::HTLCFailChannelUpdate::ChannelClosed {
+								  short_channel_id: match error_code & 0xff {
+								    8|9|10 => route.hops[next_route_hop_ix].short_channel_id, // outgoing (permanent_channel_failure/required_channel_feature_missing/unknown_next_peer)
+								    _ => route_hop.short_channel_id, // incoming channel (invalid_realm & BADONION)
+								  },
+								   is_permanent: error_code & PERM == PERM,
+								})};
+							}
+							else if error_code & UPDATE == UPDATE {
+								if err_packet.failuremsg.len() >= debug_field_size + 2 {
+									let update_len = byte_utils::slice_to_be16(&err_packet.failuremsg[debug_field_size+2..debug_field_size+4]) as usize;
+									if err_packet.failuremsg.len() >= debug_field_size + 4 + update_len {
+										if let Ok(chan_update) = msgs::ChannelUpdate::read(&mut Cursor::new(&err_packet.failuremsg[debug_field_size + 4..debug_field_size + 4 + update_len])) {
+											// if channel_update should NOT have caused the failure:
+											// MAY treat the channel_update as invalid.
+											let is_chan_update_invalid = match error_code & 0xff {
+												7 => false,
+												11 => amt_to_forward > chan_update.contents.htlc_minimum_msat,
+												12 => {
+													let new_fee = amt_to_forward.checked_mul(chan_update.contents.fee_proportional_millionths as u64).and_then(|prop_fee| { (prop_fee / 1000000).checked_add(chan_update.contents.fee_base_msat as u64) });
+													new_fee.is_none() || incoming_htlc_msat >= new_fee.unwrap() && incoming_htlc_msat >= amt_to_forward + new_fee.unwrap()
+												}
+												13 => route_hop.cltv_expiry_delta as u16 >= chan_update.contents.cltv_expiry_delta,
+												14 => false, // expiry_too_soon; always valid?
+												20 => chan_update.contents.flags & 0x0200 == 1,
+												_ => false, // unknown error code; take channel_update as valid
+											};
+											fail_channel_update = if is_chan_update_invalid { None } else {
+												Some(msgs::HTLCFailChannelUpdate::ChannelUpdateMessage {
+													msg: chan_update,
+												})
+											};
 										}
 									}
-								},
-								_c if error_code & BADONION == BADONION => {
-									//TODO
-								},
-								14 => { // expiry_too_soon
-									res = Some((None, true));
-									return;
-								}
-								_ => {
-									// node sending unknown code
-									res = Some((Some(msgs::HTLCFailChannelUpdate::NodeFailure {
-										node_id: route_hop.pubkey,
-										is_permanent: true,
-									}), false));
-									return;
 								}
 							}
+
+							res = Some((fail_channel_update, !(error_code & PERM == PERM && is_from_final_node)));
```

Comment:
If we get a None on fail_channel_update and !payment_failed (ie the data provided by the err'ing node was unreadable by us), shouldn't we default to failing something? eg if we get an UPDATE flag but couldn't parse the update, if is_chan_update_invalid, etc we should probably fail the channel, or if we get no readable flags we can fail the node wholesale.